R/plot_carp.R

```diff
@@ -290,29 +300,115 @@ carp_path_plot <- function(x,
 }
 
 #' @noRd
-#' @importFrom rlang .data
 #' @importFrom dplyr filter select summarize pull
 #' @importFrom stats as.dendrogram
-#' @importFrom dendextend set
-#' @importFrom grDevices adjustcolor
+#' @importFrom dendextend set color_branches as.ggdend
+#' @importFrom ggplot2 ggplot geom_segment scale_x_continuous scale_y_continuous theme
+#' @importFrom ggplot2 element_text xlab ylab aes element_blank labs
 carp_dendro_plot <- function(x,
                              percent,
                              k,
-                             dend.branch.width = 2,
-                             dend.labels.cex = .6,
-                             dend.ylab.cex = 1.2,
                              show_clusters = (n_args == 1L),
-                             base_colors = c("grey", "black"),
                              ...){
 
-  if(dend.branch.width <= 0){
-    crv_error(sQuote("dend.branch.width"), " must be positive.")
-  }
+  has_percent <- !missing(percent)
+  has_k       <- !missing(k)
+  n_args      <- has_percent + has_k
 
-  if(dend.labels.cex <= 0){
-    crv_error(sQuote("dend.labels.cex"), " must be positive.")
+  if(n_args > 1L){
+    crv_error("At most one of ", sQuote("percent"), " and ", sQuote("k"), " may be supplied.")
   }
 
+  d <- as.dendrogram(x)
+
+  if(show_clusters){
+
+    if(has_percent){
+      if (!is_percent_scalar(percent)) {
+        crv_error(sQuote("percent"), " must be a scalar between 0 and 1 (inclusive).")
+      }
+
+      k <- get_feature_paths(x, features = character()) %>% filter(.data$GammaPercent <= percent) %>%
+        select(.data$NCluster) %>%
+        summarize(NCluster = min(.data$NCluster)) %>%
+        pull
+    } else {
+      if (!is_integer_scalar(k)) {
+        crv_error(sQuote("k"), " must be an integer scalar (vector of length 1).")
+      }
+      if ( k <= 0 ) {
+        crv_error(sQuote("k"), " must be positive.")
+      }
+      if ( k > NROW(x$X) ) {
+        crv_error(sQuote("k"), " cannot be more than the observations in the original data set (", NROW(x$X), ").")
+      }
+
+      percent <- get_feature_paths(x, features = character()) %>% filter(.data$NCluster == k) %>%
+        select(.data$GammaPercent) %>%
+        summarize(NCluster = mean(.data$GammaPercent)) %>%
+        pull
+    }
+
+    c <- color_branches(d, k = k)
+    dend <- as.ggdend(c)
+    segs <- dend$segments
+
+    adjust <- get_feature_paths(x, features = character()) %>% filter(.data$NCluster == 1) %>%
+      select(.data$GammaPercent) %>%
+      summarize(GammaPercent = min(.data$GammaPercent)) %>%
+      pull
+    segs$y <- segs$y*adjust
+    segs$yend <- segs$yend*adjust
+
+    tree <- as.hclust(x)
+    cluster <- stats::cutree(tree, k = k)
```

Comment:
Why doing this as `stats::cutree` instead of just importing `cutree` from `stats` as we do for other functions.